### ISSUE TITLE
Methods to save space on your '/root' partition while using flatpak ... 

### ISSUE THREAD

**Issue by aiamuzz** [flatpak/flatpak#2337]
### Information
Flatpak installs at the default location `/var/lib/flatpak`. So all applications get installed inside this flatpak folder(runtimes/platform/applications ... everything in this folder) eating away at the /root partition.

**Method 1** : uses a simple hack making use of linux feature of symbolic links.
**Method 2** : uses the official flatpak feature to install applications at custom locations/paths.

### Method 1 : Unofficial 'HACK' method (this method makes sense ONLY if you have a separate partition for `/home` ... if /root & /home are on the same partition its useless to try this method)

Lets say your home is located at `/home/user` ... ('user' could be the name of your user and the same needs to be replaced with the actual user name on your system in place of `user` wherever found in the below commands)

`sudo mv /var/lib/flatpak /home/user`
`cd /var/lib`
`sudo ln -s /home/user/flatpak flatpak`

**that's it !!!** ... .... with the symbolic link at `/var/lib` ... all your flatpak applications will get routed and installed in your /home partition keeping the /root partition free.

### Method 2 : Official feature of installing it in a custom location of your choice by creating `extra.conf` file

Its best(not mandatory) if the following conditions are adhered to ...

- Its is best to have a separate partition on your internal hard disk(preferable) dedicated to installing flatpak applications.
- It is best to configure you linux OS to auto mount partition on boot ... 
- Its is best to let the partitions mount at their default locations ... some OSes mount it in `/media/user` and some mount it in `/mnt`
- It is best to name/label the partition 

The following commands are written assuming the default location being `/media/user` and partitions get auto mounted at this location ... make changes according to your own systems. ... i have a separate partition labeled it as Flatpak(name of the partition not the folder).

`sudo mkdir /media/muz/Flatpak/flatpak`
`sudo mkdir /etc/flatpak`
`sudo mkdir /etc/flatpak/installations.d`

`sudo gedit /etc/flatpak/installations.d/extra.conf`

Lets name this custom install location ...
Lets say i name it `myFlatpaks` ... you can use your own names instead, just make sure the name 'myFlatpaks' with your name in the commands/entries below.

Paste the following block in the extra.conf file and save it ...

```
[Installation "myFlatpaks"]
Path=/media/muz/Flatpak/flatpak/
DisplayName=myFlatpaks Installation
StorageType=harddisk
```

Now lets add the flathub remote to this custom install location ...

`flatpak --installation=myFlatpaks remote-add flathub https://flathub.org/repo/flathub.flatpakrepo`

Thats It !!!

**Advantages of using a Separate partition dedicated ONLY for flatpak applications.**

- In cases where the OS is reinstalled from scratch which involves formatting /root and /home partition. (Tested and Works)
- In cases where you are switching from one linux distro to another.(not yet tested as such a situation hasn't been encountered by me).

I like to call this setup of using a dedicated partition as 'INSTALL ONLY ONCE' setup !!!

### Commands to manage applications at this newly created custom install location.

**Install**
`flatpak --installation=myFlatpaks install flathub org.application.name`

**Uninstall**
`flatpak --installation=myFlatpaks uninstall flathub org.application.name`

**List**
`flatpak list`    -    to list all applications
`flatpak --installation=myFlatpaks list`     -     to list applications only from a specific installation location.

**Run**
`flatpak run flathub org.application.name`

**Comment by harishb94** [comment]
Nicely done !!

**Comment by matthiasclasen** [comment]
Why not just use --user to install in your home directory ?

**Comment by mwleeds** [comment]
I don't think this belongs as a GitHub issue. If you want to improve [the documentation](https://github.com/flatpak/flatpak-docs) around this feel free to submit a PR, but I don't think the symlink hack should go in the documentation.

**Comment by aiamuzz** [comment]
> I don't think this belongs as a GitHub issue. 

I know ... the right place would be a forum ... but as you know we do not have any dedicated forum for flatpak i raised the thread here ... @mwleeds ... if you kept this open then people could benefit from the information.

:) i know ... that is why i have mentioned it as 'Unofficial Hack'

**Comment by aiamuzz** [comment]
> Why not just use --user to install in your home directory ?

@matthiasclasen ...

I agree ... this option could easily replace the symlink method ... i did it because my Deepin OS has flatpak and their inhouse flatpak apps preinstalled out of the box ... and when i started like flatpak i found the default location was already heavy by about 3-4 GB ... so to save that space i ended up with the 'symlink' method.

a few days ago when i did a fresh install of my OS ... i uninstalled flatpak and pre-installed apps from Deepin ... did a clean install of flatpak from debian 'deb' package ... I discovered that the default location did not take up any space at all !!!

So i agree to your suggestion that --user option could just as easily replace the 'symlink' hack except for one small effort ... :) ... every time one has to enter the option `--user` ... i know i have to do a longer version of it to install to my custom location with the option `--installation=` ... honestly its extra effort to type both ...

**Would it be too difficult to create an option where we could set either `--user` , `--installation=` or the default place in `/var` as our preferred place of installation ? ... i guess if a `--system` option was created 

1. --system
2. --user
3. --installation - *

then the user could choose one to be their preferred location of installations ... and get away with the short and sweet direct `install` command with no parameters at all** 

If yes than i will start a new issue thread as a 'Feature Request : to set default install location of choice' so you can add it to your todo list ...

**Comment by mwleeds** [comment]
> as you know we do not have any dedicated forum for flatpak i raised the thread here

There's https://www.reddit.com/r/flatpak

> Would it be too difficult to create an option where we could set either --user , --installation= or the default place in /var as our preferred place of installation ?

Another way to accomplish what you want would be for the `install` command to search all installations, not just system and user, for existing remotes. So that way if you want to only use a custom installation you can have that be the only installation with remotes configured. And as is already the case if you want to only use `--user` you can have that be the only installation with remotes configured.

On the other hand that scheme is complicated a bit by the fact that GNOME Builder configures remotes in your user installation for you, but maybe that should be considered a Builder bug (cc: @chergert).

**Comment by kirbyfan64** [comment]
> as you know we do not have any dedicated forum for flatpak i raised the thread here

There's also the mailing list.

**Comment by mwleeds** [comment]
> @mwleeds ... of what you suggest above ... can i setup my system to install with the command flatpak install flathub org.qBittorrent.qbittorrent currently ... and still end up installing all my applications to 'myFlatpaks' custom location ?
> 
> If so ... i would appreciate a little more detailed procedure on how can accomplish that ...

Currently, no. I was suggesting a possible change to the install command that would be similar to what https://github.com/flatpak/flatpak/pull/2345 does. The way it could work is that if you don't use `--installation` and you have, say, flathub configured in both the system installation and your custom one, you're prompted to choose between them. If you only have the flathub remote configured in your custom installation, the custom installation would be used.

> @mwleeds ... could you update flatpak repo page with the reddit url as the flatpaks forum/place for community discussions ?

https://github.com/flatpak/flatpak/pull/2355

**Comment by aiamuzz** [comment]
> Currently, no. I was suggesting a possible change to the install command that would be similar to what #2345 does. The way it could work is that if you don't use --installation and you have, say, flathub configured in both the system installation and your custom one, you're prompted to choose between them. If you only have the flathub remote configured in your custom installation, the custom installation would be used.

oh ok you were thinking out loud on how to approach this ... :)

It may work as you state ... but wouldn't be an elegant solution in the sense that its trading off another functionality (one of NOT adding a remote) isn't it ?  ... we are trading off a valid operation of adding remotes in order to make something work for us ... this could be a interim solution but can it be permanent one ?

**Comment by aiamuzz** [comment]
@mwleeds ...

hey ... there is indeed an option by `--system` didn't know i just used it as an idea ... :)

I just noticed that the command `flatpak install flathub com.syntevo.SmartGit` and `flatpak --system install flathub com.syntevo.SmartGit` accomplish the same outcome ... both install the app to the default location.

So this command `flatpak install flathub com.syntevo.SmartGit` ... is it hardwired to use the `--system` parameter in the background ?
or

`--system` is an additional and separate option created just like `--user` ?

If its hardwired to use `--system` in the background then is it difficult to hard wire it to `--user` or `--installation=` options ?

If its hardwired to use it then wouldn't that enable you to create the above option to set a particular install location as the default ?

again ... would like to remind the above is not tech-speak ... rather a layman-speak !!! ... :)

**Comment by mwleeds** [comment]
Yes, the install command defaults to using the system installation. The way it's implemented is not quite as simple as just using the `--system` option. The difficult part about implementing this is I'm not sure where the user interface would be, and I'm not sure where the value would be stored. We have `flatpak config` but that works on a per-installation basis and you're talking about a configuration that's global across all installations and either user-wide or system-wide.

The question is, is there any good reason to have your flatpaks spread across multiple installations? If so, it could make sense to make the default installation user configurable. If not, it should be fine to just have `install` check custom installations when `--installation` is omitted, as I said above.

**Comment by aiamuzz** [comment]
> The question is, is there any good reason to have your flatpaks spread across multiple installations? If so, it could make sense to make the default installation user configurable. If not, it should be fine to just have install check custom installations when --installation is omitted, as I said above.

@mwleeds ...

I have a pretty good reason ... as you have seen me on threads here ... I have a separate partition dedicated for Flatpak ... and having setup my flatpak this way ... let me tell you that ever since i had it setup like this ... i have had to install / reinstall my OS more than 3 time now ... and i hadn't to break a sweat to bother about my flatpak apps on this partition ... all i did was setup the config file for pointing to this dedicated partition and voila ... I have been trying out 62 flatpak apps installed there  ... and all work like a charm without breaking a sweat !!!

Just like how Flatpak's philosophy for devs has been one of ... Develop-once-distribute-on-all-distro's ... I have been able to pull off a ... Install-once-and-forget-about-it philosophy ... i am yet to test to see if this setup chimes on a completely random flavor of linux ... but on the same distro its been working like a charm.

So i would say ... 52 app installations with the additional parameters ... with a default Install location mechanism i would say i would save a lot of RSI to my poor fingers :)

**Comment by aiamuzz** [comment]
@mwleeds  ...

not to forget space in the `/root` partition is always at a premium ... so installing either at a separate `/home` partition or as in my case an all together separate dedicated partition makes all the sense ... don't you think ?

having our flatpak applications spread across multiple installations makes perfect sense for saving space and the reason described above !

**Comment by chergert** [comment]
Just a note on the Builder side of things. We use `--user` because we automatically install SDKs when they are missing from the users installation. If we were to use `--system`, then we'd be requiring polkit authorization without much context to the user of why it's being displayed. For that reason, I'm hesitant to use anything other than `--user` or possible another private installation.

What I don't like about the private installation, is that we'd need a way to promote an SDK/Runtime to the system so the user doesn't have to download it 2x.

**Comment by aiamuzz** [comment]
> and I'm not sure where the value would be stored

how about here ...

```
[Installation "myFlatpaks"]
Path=/media/muz/Flatpak/flatpak/
DisplayName=myFlatpaks Installation
StorageType=harddisk
```

something like ...

```
[Installation "myFlatpaks"]
Path=/media/muz/Flatpak/flatpak/
DisplayName=myFlatpaks Installation
StorageType=harddisk
**SetDefaultInstall=yes**
```
this flag could determine which argument the `install` could default to ...

now the problem arises on how/where to save information for the --user argument ...

How about a similar config file ... say `user.config`

```
Path=/home/muz/flatpak/
SetDefaultInstall=no
```

and another config for --system ... say `system.config`

```
Path=/var/lib/flatpak/
SetDefaultInstall=no
```

or all the above clubbed into a single config file ... with the `install` command grabbing it from this file the info on where to install apps ...

**Comment by aiamuzz** [comment]
> The question is, is there any good reason to have your flatpaks spread across multiple installations?

some more reasons to this question ...

**Gamers**
as flatpak matures over time and more & more apps get added especially those of Games ... games again are usually storage-heavy ... it would help gamer's to install all games in one dedicated partition.

**Dev's**
similarly ... devs would may be like to keep all their tools in one place(and flathub has quite a few dev tools already added to it) ... on their own dedicated space.

I hope you guys realize that dedicated private installation on dedicated spaces(partitions) are almost like 'appimage's' ... bringing its  'run anywhere'  feature to flatpak too ...

**Comment by mwleeds** [comment]
> Just a note on the Builder side of things. We use --user because we automatically install SDKs when they are missing from the users installation. If we were to use --system, then we'd be requiring polkit authorization without much context to the user of why it's being displayed. For that reason, I'm hesitant to use anything other than --user or possible another private installation.

Installing an app or runtime from an existing remote doesn't require polkit authorization. I believe you'd only have to ask for authorization if they didn't have the flathub remote already configured (or if they're not a sudoer). Anyway, I noticed [this issue exists](https://gitlab.gnome.org/GNOME/gnome-builder/issues/501) so I can comment there.

**Comment by mwleeds** [comment]
@aiamuzz I think you misunderstood the question I was asking. I'm not questioning the value of using custom installations to move flatpaks to another partition. I'm asking if you need to **also** have flatpaks in the system installation when you're primarily using the custom installation. It looks like you have five for example which I guess were installed by your distribution.

**Comment by aiamuzz** [comment]
> I'm asking if you need to also have flatpaks in the system installation when you're primarily using the custom installation. It looks like you have five for example which I guess were installed by your distribution.

@mwleeds ... ok i get it ... advantages of custom installation aside ... 

The need to have flatpaks in places besides custom installations could by many ...

- if like i said the custom installation where maintained as per category ... for eg 'gaming' apps or 'dev' apps ... it could be any criteria depending on the user.
- if the private installations were running out of space, then i could use `--user` or `--system` 
- say if there are apps that i would like to try it ... i could use `--user` or `--system` as temporary places for such purposes ... and if i grow a liking to it then i could move it to my private installations.

these are some of the reasons i could think of where the approach you-where-thinking-out-load earlier of adding the 'remote' only to the installation you want the install command to default would kind of play restrictive ... taking away the users freedom of multiple install locations we currently have ...

**Comment by aiamuzz** [comment]
@gusbemacbe ... could you post the outputs of the following commands ?

`cd into_the directory_where_the_flatpak_folder_is_present`
`ls -ltr`
`cat /etc/flatpak/installations.d/extra.conf`

**Comment by gusbemacbe** [comment]
Hi @aiamuzz 

Sorry, I went to bed.

Here are:

```bash
$ cd '/run/media/gusbemacbe/Gustavo Backup/'
$ cd Applications/flatpak
$ ls -ltr
total 128
drwxrwxrwx 1 gusbemacbe 131072 Jan 11 05:16 repo

$  cat /etc/flatpak/installations.d/extra.conf
[Installation "extra"]
Path=/run/media/gusbemacbe/Gustavo\ Backup/Applications/flatpak/
DisplayName=Extra Installation
StorageType=harddisk
```

**Comment by PhoneixS** [comment]
> If yes than i will start a new issue thread as a 'Feature Request : to set default install location of choice' so you can add it to your todo list ...

I think there is some interest in this, Did you open the request? I can't find it.

**Comment by Yoannjap** [comment]
> ### Information
> 
> Flatpak installs at the default location `/var/lib/flatpak`. So all applications get installed inside this flatpak folder(runtimes/platform/applications ... everything in this folder) eating away at the /root partition.
> 
> **Method 1** : uses a simple hack making use of linux feature of symbolic links. **Method 2** : uses the official flatpak feature to install applications at custom locations/paths.
> ### Method 1 : Unofficial 'HACK' method (this method makes sense ONLY if you have a separate partition for `/home` ... if /root & /home are on the same partition its useless to try this method)
> 
> Lets say your home is located at `/home/user` ... ('user' could be the name of your user and the same needs to be replaced with the actual user name on your system in place of `user` wherever found in the below commands)
> 
> `sudo mv /var/lib/flatpak /home/user` `cd /var/lib` `sudo ln -s /home/user/flatpak flatpak`
> 
> **that's it !!!** ... .... with the symbolic link at `/var/lib` ... all your flatpak applications will get routed and installed in your /home partition keeping the /root partition free.
> ### Method 2 : Official feature of installing it in a custom location of your choice by creating `extra.conf` file
> 
> Its best(not mandatory) if the following conditions are adhered to ...
> 
>     * Its is best to have a separate partition on your internal hard disk(preferable) dedicated to installing flatpak applications.
> 
>     * It is best to configure you linux OS to auto mount partition on boot ...
> 
>     * Its is best to let the partitions mount at their default locations ... some OSes mount it in `/media/user` and some mount it in `/mnt`
> 
>     * It is best to name/label the partition
> 
> 
> The following commands are written assuming the default location being `/media/user` and partitions get auto mounted at this location ... make changes according to your own systems. ... i have a separate partition labeled it as Flatpak(name of the partition not the folder).
> 
> `sudo mkdir /media/muz/Flatpak/flatpak` `sudo mkdir /etc/flatpak` `sudo mkdir /etc/flatpak/installations.d`
> 
> `sudo gedit /etc/flatpak/installations.d/extra.conf`
> 
> Lets name this custom install location ... Lets say i name it `myFlatpaks` ... you can use your own names instead, just make sure the name 'myFlatpaks' with your name in the commands/entries below.
> 
> Paste the following block in the extra.conf file and save it ...
> 
> ```
> [Installation "myFlatpaks"]
> Path=/media/muz/Flatpak/flatpak/
> DisplayName=myFlatpaks Installation
> StorageType=harddisk
> ```
> 
> Now lets add the flathub remote to this custom install location ...
> 
> `flatpak --installation=myFlatpaks remote-add flathub https://flathub.org/repo/flathub.flatpakrepo`
> 
> Thats It !!!
> 
> **Advantages of using a Separate partition dedicated ONLY for flatpak applications.**
> 
>     * In cases where the OS is reinstalled from scratch which involves formatting /root and /home partition. (Tested and Works)
> 
>     * In cases where you are switching from one linux distro to another.(not yet tested as such a situation hasn't been encountered by me).
> 
> 
> I like to call this setup of using a dedicated partition as 'INSTALL ONLY ONCE' setup !!!
> ### Commands to manage applications at this newly created custom install location.
> 
> **Install** `flatpak --installation=myFlatpaks install flathub org.application.name`
> 
> **Uninstall** `flatpak --installation=myFlatpaks uninstall flathub org.application.name`
> 
> **List** `flatpak list` - to list all applications `flatpak --installation=myFlatpaks list` - to list applications only from a specific installation location.
> 
> **Run** `flatpak run flathub org.application.name`

Thanks a lot aimauzz for your brillant method !!
I have 2 separated / and /home partitions, so I used the very easy 1st method.

2 questions then :
- If someday I want to upgrade Ubuntu from a liveCD (only formatting / partition) do I have to enter your moving flatpak files commands again, or something else, in order to recover my flatpak programs' media, files and settings?
- If I prefer to upgrade from my current distribution, using do-release-upgrade -d command, do I have to enter your flatpak commands again, or something esle?

Thanks for your help!

Yoannjap
